examples/python/gui_demo/app_context.py

```diff
@@ -86,28 +86,36 @@ def load_icons(self, directory):
             images[file.split('.')[0]] = image
         self.icons = images
 
-    def short_id(self, global_id: str):
+    def short_id(self, global_id):
+        if not global_id or not isinstance(global_id, str):
+            return ''
+
         parts = global_id.split('/')
-        n_parts = len(parts)
-        devices = ['']
-        if n_parts > 3: # empty + device + Sig + ID
-            if parts[-2] == 'Sig':
-                signal = parts[-1]
-                if signal:
-                    for index, part in enumerate(parts[:-2]):
-                        if part == 'Dev' and index + 1 < n_parts:
-                            devices.append(parts[index + 1])
-                    return '/'.join(devices) + '/' + signal
-        return ''
+        filtered_parts = []
+
+        for index, part in reversed(list(enumerate(parts))):
+            # iterate until device and skip root device
+            # 0: empty; 1: root device id;
+            if part == 'Dev' or index < 2:
+                break
+            # filter other default parts
+            if part not in ('IO', 'FB', 'Sig'):
+                filtered_parts.append(part)
+        return '/'.join(reversed(filtered_parts))
 
     def update_signals_for_device(self, device):
         if device is None:
             return
+
         self.signals[device.global_id] = {}
         for signal in device.signals_recursive:
-            self.signals[device.global_id][self.short_id(
-                signal.global_id)] = signal
-            
+            short_id = self.short_id(signal.global_id)
+
+            if short_id not in self.signals[device.global_id]:
+                self.signals[device.global_id][short_id] = signal
+            else: #insert global id in case of collision
+                self.signals[device.global_id][signal.global_id] = signal
```

Comment:
If we have a collision, we should also retroactively replace the one we clashed with.

Reply:
Done